compiler/src/dotty/tools/dotc/typer/CrossVersionChecks.scala

```diff
@@ -52,18 +52,22 @@ class CrossVersionChecks extends MiniPhase:
       owner.isDeprecated
       || isEnumOwner(owner)
 
-    /**Scan the chain of outer declaring scopes from the current context
+    /**Skip warnings for synthetic members of case classes during declaration and
+     * scan the chain of outer declaring scopes from the current context
      * a deprecation warning will be skipped if one the following holds
      * for a given declaring scope:
      * - the symbol associated with the scope is also deprecated.
      * - if and only if `sym` is an enum case, the scope is either
      *   a module that declares `sym`, or the companion class of the
      *   module that declares `sym`.
      */
-    def skipWarning(using Context) =
-      ctx.owner.ownersIterator.exists(if sym.isEnumCase then isDeprecatedOrEnum else _.isDeprecated)
+    def skipWarning(using Context): Boolean =
+      (ctx.owner.is(Synthetic) && sym.is(CaseClass)) ||
+        ctx.owner.ownersIterator.exists(if sym.isEnumCase then isDeprecatedOrEnum else _.isDeprecated)
 
-    for annot <- sym.getAnnotation(defn.DeprecatedAnnot) do
+    // Also check for deprecation of the companion class for synthetic methods
+    val toCheck: List[Symbol] = sym :: Option.when(sym.isAllOf(SyntheticMethod))(sym.owner.companionClass).toList
```

Comment:
```suggestion
    val toCheck = sym :: (if sym.isAllOf(SyntheticMethod) then sym.owner.companionClass :: Nil else Nil)
```
This is faster and clearer.